pkg/readiness/objset.go

```diff
@@ -27,9 +27,35 @@ type objKey struct {
 	namespacedName types.NamespacedName
 }
 
-type objSet map[objKey]struct{}
-
 func (k objKey) String() string {
 	return fmt.Sprintf("%s [%s]", k.namespacedName.String(), k.gvk.String())
+}
+
+type objData struct {
+	retries int
+}
 
+// decrementRetries handles objData retries, and returns `true` if it's time to delete the objData entry
+func (o *objData) decrementRetries() bool {
+	// if retries is less than 0, allowed retries are infinite
+	if o.retries < 0 {
+		return false
+	}
+
+	// If we have retries left, use one
+	if o.retries > 0 {
+		o.retries--
+		return false
+	}
+
+	// if we have zero retries, we can delete
+	return true
 }
+
+type objDataFactory func() *objData
+
+func objDataFromFlags() *objData {
+	return &objData{retries: *readinessRetries}
+}
+
+type objSet map[objKey]objData
```

Comment:
Correct me if I'm wrong, but only the `expect` set uses counters. I'm finding it confusing reading the code and checking whether `Observe` is still idempotent because it overwrites a counter nobody is looking at. Should we create a separate type for `expect` and all the rest?

Reply:
+1

Reply:
That sounds good to me.

I've added a second type: `objRetrySet`.  This type is used only for `t.expect`, while `t.seen` and others will be assigned like:

```
t.seen[k] = struct{}{}
```

as they were done before.  I believe makes the distinction clear in code.

Reply:
Per this [other comment](https://github.com/open-policy-agent/gatekeeper/pull/1014#discussion_r549541068), I've now moved the retries tracking into its own map.

So, `t.expect` is back to using the original `objSet`.